README.md
# A nice boilerplate for a new AWS Lambda function

[![Build Status](https://travis-ci.org/5orenso/aws-lambda-boilerplate.svg?branch=master)](https://travis-ci.org/5orenso/aws-lambda-boilerplate)
[![Coverage Status](https://coveralls.io/repos/github/5orenso/aws-lambda-boilerplate/badge.svg?branch=master)](https://coveralls.io/github/5orenso/aws-lambda-boilerplate?branch=master)

A good start for an AWS Lambda function. If you combine this with the [aws-lambda-autodeploy-lambda](https://github.com/5orenso/aws-lambda-autodeploy-lambda) you will have a very fast deployment process. When you push code to Github.com it will only take a couple of minutes to run all your test, deploy you code to S3 and refresh the Lambda function. All automatic while you're drinking a fresh cup of newly brewed coffee.

__Features:__
* Boilerplate for [AWS Lamda](https://aws.amazon.com/lambda/) functions.
* Automated unit tests with [NPM](https://www.npmjs.com/) and [Jest](https://jestjs.io/).
* Code coverage with [Jest](https://jestjs.io/).
* Code style and hits with [ESLint](https://eslint.org/).
* CI integration with [Travis](https://travis-ci.org/).
* And last but not least; deployment with [AWS CloudFormation](https://aws.amazon.com/cloudformation/)


## Getting started with development

```
$ git clone
$ npm install
$ cp ./bin/set-env-dist.sh ./bin/set-env.sh
$ . ./bin/set-env.sh
$ npm run test:watch
```


### Run all tests

```bash
$ npm test
```


### Lint all of the relevant code

```bash
$ npm run lint
```


### Build a new release

```bash
$ npm run build
```


### Report coverage to Coveralls.io

```bash
$ npm run coverage
```


### Update changelog

```bash
$ bash ./changelog.sh
```


## Travis and Coveralls setup

To get auto coveralls reporting working you need to add your encrypted Coveralls token in `.travis.yml`.

This is how you do it:
```bash
$ travis encrypt COVERALLS_REPO_TOKEN=<super_secret coveralls token> --add env.global
```


### Steps for preparing your repo:

1. Connect your Github.com repo to your Travis CI account. https://travis-ci.org/
2. Add `AWS_ACCESS_KEY` and `AWS_SECRET_KEY` to your Travis CI environment.
3. Edit `.travis.yml` to match your `bucket`, `upload-dir` and `region`.
4. Copy `.coveralls-dist.yml` to `.coveralls.yml` and add you secret key.


Next time you push to your master brach Travis will run your tests and if all is green report coverage to Coveralls.

Isn't this great? :D


### CloudFormation

```bash
# To create a new stack:
$ bash ./bin/create-stack.sh

# To update a current stack:
$ bash ./bin/create-stack.sh -u
```


### Howto check for vulnerabilities in modules

```bash
# Install Node Security Platform CLI
$ npm install nsp --global  

# From inside your project directory
$ nsp check  
```


### Howto upgrade modules

```bash
$ npm install -g npm-check-updates
$ ncu -u
$ npm install --save --no-optional
```


## Other Resources

* [AWS Basic setup with Cloudformation](https://github.com/5orenso/aws-cloudformation-base)
* [AWS Lambda boilerplate](https://github.com/5orenso/aws-lambda-boilerplate)
* [Automated AWS Lambda update](https://github.com/5orenso/aws-lambda-autodeploy-lambda)
* [AWS API Gateway setup with Cloudformation](https://github.com/5orenso/aws-cloudformation-api-gateway)
* [AWS IoT setup with Cloudformation](https://github.com/5orenso/aws-cloudformation-iot)


CHANGELOG.md
* _2019-01-18_: Updated docs. (HEAD -> master, origin/master, origin/HEAD) [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/86d83b3e32f05028bdc0678030514453804a242f">view commit</a>
* _2019-01-18_: Fixing coveralls integration with Travis. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/d19a7f5820d924502885c9b144e46448dfa3e3cc">view commit</a>
* _2019-01-18_: Added reporting to coveralls.io [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/d30182ef95ffb72274a405eddada7cc48d1fb971">view commit</a>
* _2019-01-17_: Removed deploy from travis.yml [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/5389989825806610b59e77a2617677bdc2a0be8f">view commit</a>
* _2019-01-17_: Removed error [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/4289bed92e70dea120faa62a7ba86c222ee0ba6a">view commit</a>
* _2019-01-17_: Updated shit to get rid of grunt and other old stuff. Added Jest and Husky [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/8aa1aa52800a8eeb6d08bf3ba9692bde9b7f15d6">view commit</a>
* _2019-01-14_: Updated docs [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/cc4306d6967ec1ca65bb3749e99dbcf1b7d08e2e">view commit</a>
* _2019-01-14_: New package-lock.json [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/3fdf645571623af28d76cec3f5ff6c541467bcf4">view commit</a>
* _2019-01-14_: Upgraded libs. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/b7dfb2f9c43b9a0ce28bf31fbc85e5aadecbf28b">view commit</a>
* _2019-01-14_: Fixed Travis settings. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/9d8076c41dc219e403e06da219f16d7741547433">view commit</a>
* _2019-01-14_: Updated to use ES6 and Eslint. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/0f02a781fa9a229b15b6ea7dd20c0fc8ad84ff1f">view commit</a>
* _2018-09-04_: More resources. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/4de33c95d1c084e49619f0dc40d7337aa0c86ec3">view commit</a>
* _2017-08-29_: More info... [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/f8607a7fd14e2345c0352cee5412f2ff1f50b133">view commit</a>
* _2017-08-29_: Added autodeply feature. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/d13600392853c05b96082d8b266a78adf62f2dd9">view commit</a>
* _2017-05-04_: Added code coverage report script. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/19872c1b4a15154b22632a250b4117e71b1bf9e9">view commit</a>
* _2017-05-04_: Added install script. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/0ed1af71a730a8a7e0b624f25b9e41f17e540b2f">view commit</a>
* _2017-05-04_: Added changelog script. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/376e53de7a0123b002e872c3095c873c6fd8a47c">view commit</a>
* _2017-01-23_: Added resources. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/b25d2c6001b0a6545815054417ddd4bee6a00ed9">view commit</a>
* _2017-01-23_: Refactored to be a valid AWS Gateway reponse without any libs. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/267d396c8faca1d25e329704daf5e949cd7fd901">view commit</a>
* _2017-01-20_: Added test of branch name. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/e1ac30d1580ef8682b41bdd59e1a4d940c385b13">view commit</a>
* _2017-01-20_: Merge pull request #1 from 5orenso/test [Øistein Sørensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/acd3c7af537e9d3e34016c8620a614f59d6ae254">view commit</a>
* _2017-01-20_: Trying to fix branch name. (origin/test, test) [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/6cf6ad2e86a581b33b03fffc527453b67ef10efb">view commit</a>
* _2017-01-20_: Trying to fix branch name. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/010609198da7bebc29fd7cd8d62e55a57c93e275">view commit</a>
* _2017-01-20_: Trying to fix branch name. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/87443545cd0fd4c7d1079b8d7eee3071e678411e">view commit</a>
* _2017-01-20_: Trying to fix branch name. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/82fb18e711b34148d992dd7ae0fc435c1af389b2">view commit</a>
* _2017-01-20_: Trying to fix branch name. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/c396d22287cc2d9e9fee3e2a2f541307b4f853b8">view commit</a>
* _2017-01-20_: Fixed travis settings. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/9101ed5c3252b13fcdf033f4a2f385587589d06c">view commit</a>
* _2017-01-20_: Added deployment for all branches [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/8a99b7f01016847c4b9736eba8c9efa3593d8b63">view commit</a>
* _2017-01-20_: Fixed jshint issues. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/c765dc555ece401f1e46f35eba85c4862771e5ca">view commit</a>
* _2017-01-20_: Updated docs [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/05129fc31714ac91f671194c81d06cf6535411dd">view commit</a>
* _2017-01-20_: Added timing [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/2ba4a25e37b2b7d11f64bf99576f9c88367699a6">view commit</a>
* _2017-01-20_: Added timing and branch builds. [Oistein Sorensen] <a href="http://github.com/5orenso/aws-lambda-boilerplate/commit/f45190305dbe4bc658781e6c6cd325735145f7d8">view commit</a>

test/buster.js
const config = module.exports;

config['AWS Lambda'] = {
    environment: 'node',
    rootPath: '../',
    tests: [
        'test/**/*-test.js',
    ],
    'buster-istanbul': {
        outputDirectory: 'coverage',
        format: 'lcov',
    },
    sources: [
        '*.js',
        '!Gruntfile.js',
    ],
    extensions: [
        require('buster-istanbul'), // eslint-disable-line
    ],
};


test/index-test.js
'use strict';

const buster = require('buster');
const index = require('../index.js');

const assert = buster.assert;

// var AWS = require('aws-sdk-mock');
// AWS.mock('S3', 'getObject', function (params, callback) {
//     if (params.Bucket === 'boom') {
//         callback('BOOM!');
//     }
//     callback(undefined, myFile);
// });

// jscs:disable
const apiGwLambdaRequest = {
    resource: '/hello-world',
    path: '/hello-world/',
    httpMethod: 'GET',
    headers: {
        Accept: '*/*',
        'Accept-Encoding': 'gzip, deflate, sdch, br',
        'Accept-Language': 'en-US,en;q=0.8,da;q=0.6,nb;q=0.4,sv;q=0.2',
        'Cache-Control': 'no-cache',
        'CloudFront-Forwarded-Proto': 'https',
        'CloudFront-Is-Desktop-Viewer': 'true',
        'CloudFront-Is-Mobile-Viewer': 'false',
        'CloudFront-Is-SmartTV-Viewer': 'false',
        'CloudFront-Is-Tablet-Viewer': 'false',
        'CloudFront-Viewer-Country': 'NO',
        'Content-Type': 'text/plain',
        DNT: '1',
        Host: 'my-api-id.execute-api.eu-west-1.amazonaws.com',
        'Postman-Token': '711128bf-f39a-34bc-b86f-5e482365e02b',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) '
            + 'Chrome/55.0.2883.95 Safari/537.36',
        Via: '1.1 asdf9sdf9sadf9asdf9asdf9asdf9sadf9asdf9.cloudfront.net (CloudFront)',
        'X-Amz-Cf-Id': 'asdlkfjasdfasdfasdfasdfAsdfASDFASDFasdfasdfasdfasdfasdf==',
        'X-Forwarded-For': '127.0.0.1, 10.0.0.1',
        'X-Forwarded-Port': '443',
        'X-Forwarded-Proto': 'https',
    },
    queryStringParameters: {
        q: 'test',
    },
    pathParameters: null,
    stageVariables: null,
    requestContext: {
        accountId: '100000000000',
        resourceId: 'aaaabbbb',
        stage: 'dev',
        requestId: 'asdfasdf-asdfasdfasdf-asdfadsfasdf-asdfasdf',
        identity: {
            cognitoIdentityPoolId: null,
            accountId: null,
            cognitoIdentityId: null,
            caller: null,
            apiKey: null,
            sourceIp: '127.0.0.1',
            accessKey: null,
            cognitoAuthenticationType: null,
            cognitoAuthenticationProvider: null,
            userArn: null,
            userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) '
                + 'Chrome/55.0.2883.95 Safari/537.36',
            user: null,
        },
        resourcePath: '/hello-world',
        httpMethod: 'GET',
        apiId: 'my-api-id',
    },
    body: null,
    isBase64Encoded: false,
};

const apiGwLambdaResponse = {
    statusCode: 200,
    body: '{"q":"test"}',
    headers: { 'Content-Type': 'application/json' },
};

const apiGwLambdaBadRequstResponse = {
    body: 'Unsupported method "BADREQUEST"',
    headers: { 'Content-Type': 'application/json' },
    statusCode: 400,
};
// jscs:enable

buster.testCase('index.js lambda handler', {
    setUp: () => {
        this.timeout = 2000;
    },
    tearDown: () => {
    },
    'GET test cases': {
        'Should return querystring input in the correct format': (done) => {
            index.handler(apiGwLambdaRequest, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'POST test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'POST',
                requestContext: {
                    httpMethod: 'POST',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'PUT test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'PUT',
                requestContext: {
                    httpMethod: 'PUT',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'PATCH test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'PATCH',
                requestContext: {
                    httpMethod: 'PATCH',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },

    'DELETE test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'DELETE',
                requestContext: {
                    httpMethod: 'DELETE',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'COPY test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'COPY',
                requestContext: {
                    httpMethod: 'COPY',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'HEAD test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'HEAD',
                requestContext: {
                    httpMethod: 'HEAD',
                },
            });
            const response = Object.assign({}, apiGwLambdaResponse, {
                body: '{}',
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, response);
                done();
            });
        },
    },
    'OPTIONS test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'OPTIONS',
                requestContext: {
                    httpMethod: 'OPTIONS',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaResponse);
                done();
            });
        },
    },
    'BADREQUEST test cases': {
        'Should return querystring input in the correct format': (done) => {
            const request = Object.assign({}, apiGwLambdaRequest, {
                httpMethod: 'BADREQUEST',
                requestContext: {
                    httpMethod: 'BADREQUEST',
                },
                body: '{"q":"test"}',
                queryStringParameters: {},
            });
            index.handler(request, undefined, (error, result) => {
                assert.equals(result, apiGwLambdaBadRequstResponse);
                done();
            });
        },
    },
});


__test__/index.test.js
const index = require('../index.js');

const apiGwLambdaRequest = {
    resource: '/hello-world',
    path: '/hello-world/',
    httpMethod: 'GET',
    headers: {
        Accept: '*/*',
        'Accept-Encoding': 'gzip, deflate, sdch, br',
        'Accept-Language': 'en-US,en;q=0.8,da;q=0.6,nb;q=0.4,sv;q=0.2',
        'Cache-Control': 'no-cache',
        'CloudFront-Forwarded-Proto': 'https',
        'CloudFront-Is-Desktop-Viewer': 'true',
        'CloudFront-Is-Mobile-Viewer': 'false',
        'CloudFront-Is-SmartTV-Viewer': 'false',
        'CloudFront-Is-Tablet-Viewer': 'false',
        'CloudFront-Viewer-Country': 'NO',
        'Content-Type': 'text/plain',
        DNT: '1',
        Host: 'my-api-id.execute-api.eu-west-1.amazonaws.com',
        'Postman-Token': '711128bf-f39a-34bc-b86f-5e482365e02b',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) '
            + 'Chrome/55.0.2883.95 Safari/537.36',
        Via: '1.1 asdf9sdf9sadf9asdf9asdf9asdf9sadf9asdf9.cloudfront.net (CloudFront)',
        'X-Amz-Cf-Id': 'asdlkfjasdfasdfasdfasdfAsdfASDFASDFasdfasdfasdfasdfasdf==',
        'X-Forwarded-For': '127.0.0.1, 10.0.0.1',
        'X-Forwarded-Port': '443',
        'X-Forwarded-Proto': 'https',
    },
    queryStringParameters: {
        q: 'test',
    },
    pathParameters: null,
    stageVariables: null,
    requestContext: {
        accountId: '100000000000',
        resourceId: 'aaaabbbb',
        stage: 'dev',
        requestId: 'asdfasdf-asdfasdfasdf-asdfadsfasdf-asdfasdf',
        identity: {
            cognitoIdentityPoolId: null,
            accountId: null,
            cognitoIdentityId: null,
            caller: null,
            apiKey: null,
            sourceIp: '127.0.0.1',
            accessKey: null,
            cognitoAuthenticationType: null,
            cognitoAuthenticationProvider: null,
            userArn: null,
            userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) '
                + 'Chrome/55.0.2883.95 Safari/537.36',
            user: null,
        },
        resourcePath: '/hello-world',
        httpMethod: 'GET',
        apiId: 'my-api-id',
    },
    body: null,
    isBase64Encoded: false,
};

const apiGwLambdaResponse = {
    statusCode: 200,
    body: '{"q":"test"}',
    headers: { 'Content-Type': 'application/json' },
};

const apiGwLambdaBadRequstResponse = {
    body: 'Unsupported method "BADREQUEST"',
    headers: { 'Content-Type': 'application/json' },
    statusCode: 400,
};

test('adds 1 + 2 to equal 3', () => {
    expect(1 + 2).toBe(3);
});

test('GET test cases: Should return querystring input in the correct format', () => {
    index.handler(apiGwLambdaRequest)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('POST test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'POST',
        requestContext: {
            httpMethod: 'POST',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('PUT test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'PUT',
        requestContext: {
            httpMethod: 'PUT',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('PATCH test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'PATCH',
        requestContext: {
            httpMethod: 'PATCH',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('DELETE test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'DELETE',
        requestContext: {
            httpMethod: 'DELETE',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('COPY test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'COPY',
        requestContext: {
            httpMethod: 'COPY',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('HEAD test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'HEAD',
        requestContext: {
            httpMethod: 'HEAD',
        },
    });
    const response = Object.assign({}, apiGwLambdaResponse, {
        body: '{}',
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(response);
        });
});

test('OPTIONS test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'OPTIONS',
        requestContext: {
            httpMethod: 'OPTIONS',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaResponse);
        });
});

test('BADREQUEST test cases: Should return querystring input in the correct format', () => {
    const request = Object.assign({}, apiGwLambdaRequest, {
        httpMethod: 'BADREQUEST',
        requestContext: {
            httpMethod: 'BADREQUEST',
        },
        body: '{"q":"test"}',
        queryStringParameters: {},
    });
    index.handler(request)
        .then((result) => {
            expect(result).toEqual(apiGwLambdaBadRequstResponse);
        });
});


install.sh

#!/bin/bash

echo 'You should probably do this first:'
echo '1. Generate a new repo on github.com.'
echo '2. Clone repo to you project home dir.'

echo '3. Enter path of new repo without trailing slash (ie: "../my-new-web-repo"): '
read PATH

# Copying paths
/bin/cp -Rv bin $PATH/.
/bin/cp -Rv cloudformation $PATH/.
/bin/cp -Rv __test__ $PATH/.

# Copying files
/bin/cp -v .coveralls-dist.yml $PATH/.
/bin/cp -v .eslintrc.json $PATH/.
/bin/cp -v .gitignore $PATH/.
/bin/cp -v .travis.yml $PATH/.
/bin/cp -v index.js $PATH/.
/bin/cp -v install.sh $PATH/.
/bin/cp -v LICENSE $PATH/.
/bin/cp -v package.json $PATH/.
/bin/cp -v README.md $PATH/.
/bin/cp -v changelog.sh $PATH/.

cd $PATH
/bin/cp -v ./.coveralls-dist.yml ./.coveralls.yml

echo ''
echo 'Todo:'
echo '1. Update coverlalls.yml with the correct token.'
echo '2. Update ./bin/set-env.sh'
echo '3. Add coveralls token to .travis.yml. See README.md for details.'
echo '3. Run `npm install`'
echo '4. Run your tests `npm test`.'
echo ''
echo 'All set :)'
echo ''


Gruntfile.js
'use strict';

module.exports = function gruntConfig(grunt) {
    require('time-grunt')(grunt); // eslint-disable-line
    let currentBranchName = (process.env.TRAVIS_PULL_REQUEST_BRANCH === '' ?
        process.env.TRAVIS_BRANCH : process.env.TRAVIS_PULL_REQUEST_BRANCH);
    if (currentBranchName === 'master' || currentBranchName == null) {
        currentBranchName = '';
    } else {
        currentBranchName = `-${currentBranchName}`;
    }
    // Project configuration.
    grunt.initConfig({
        currentBranch: currentBranchName,
        packetName: 'aws-lambda-boilerplate',
        staticLambdaBucket: 'my-lambda-deploy-test-bucket',
        autoLambdaBucket: 'my-lambda-autodeploy-lambda-bucket',
        pkg: grunt.file.readJSON('package.json'),
        eslint: {
            options: {
                config: '.eslintrc.json',
                reset: true,
            },
            target: ['*.js', 'test/**/*.js', 'lib/**/*.js'],
        },
        watch: {
            all: {
                files: ['test/**/*.js', '*.js', 'lib/**/*.js'],
                tasks: ['lint', 'buster:unit'],
            },
        },
        buster: {
            unit: {},
        },
        shell: {
            multiple: {
                command: [
                    'mkdir -p dist',
                    'mv ./node_modules ./node_modules2',
                    'npm install --production',
                    'zip -FSr  dist/<%= packetName %><%= currentBranch %>.zip ./ ' +
                        '-x "*dist*" "bin/*" ".git*" "*.md" "*.DS_Store" "*.sh" "*test*" ' +
                        '"package.json" "Gruntfile.js" ' +
                        '"*node_modules2*" "*coverage/*" ".js*" "*doc/*"',
                    'rm -rf node_modules',
                    'mv ./node_modules2 ./node_modules',
                    'echo ""',
                    'echo "TODO:"',
                    'echo "First time: Upload the zip file to S3 to be able to run it from Lambda."',
                    'echo "$ aws s3 cp dist/<%= packetName %><%= currentBranch %>.zip s3://<%= staticLambdaBucket %>/'
                        + '<%= packetName %><%= currentBranch %>.zip"',
                    'echo ""',
                    'echo "Next time: Upload the zip file to S3 to have the Lambda function automatically updated."',
                    'echo "$ aws s3 cp dist/<%= packetName %><%= currentBranch %>.zip s3://<%= autoLambdaBucket %>/'
                        + '<%= packetName %><%= currentBranch %>.zip"',
                ].join('&&'),
            },
        },
        coveralls: {
            realCoverage: {
                src: 'coverage/lcov.info',
            },
        },
    });

    // These plugins provide necessary tasks.
    grunt.loadNpmTasks('grunt-contrib-nodeunit');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-shell');
    grunt.loadNpmTasks('grunt-buster');
    grunt.loadNpmTasks('grunt-coveralls');
    grunt.loadNpmTasks('grunt-eslint');

    // Default task.
    grunt.registerTask('es', ['eslint']);
    grunt.registerTask('lint', ['eslint']);
    grunt.registerTask('default', ['lint', 'buster:unit', 'coveralls:realCoverage']);
    grunt.registerTask('coverage', ['coveralls:realCoverage']);
    grunt.registerTask('test', 'buster:unit');
    grunt.registerTask('check', ['watch']);
    grunt.registerTask('artifact', ['shell']);
    grunt.registerTask('build', ['artifact']);
};


index.js
'use strict';

/**
* @fileOverview Lambda handler API Gateway requests.
* @name Lambda handler
*/

/**
* Lambda handler.
* @param {object} event - Object passed on from AWS Lambda.
* @param {object} context - Object passed on from AWS Lambda.
* @param {function} callback - Should be called when done..
*/
exports.handler = async (event) => {
    // console.log('Received event:', JSON.stringify(event, null, 2));
    const done = (err, res) => ({
        statusCode: err ? 400 : 200,
        body: err ? err.message : JSON.stringify(res),
        headers: {
            'Content-Type': 'application/json',
        },
    });

    switch (event.httpMethod) {
        case 'GET':
            return done(null, event.queryStringParameters);
        case 'POST':
            return done(null, JSON.parse(event.body));
        case 'PUT':
            return done(null, JSON.parse(event.body));
        case 'PATCH':
            return done(null, JSON.parse(event.body));
        case 'DELETE':
            return done(null, JSON.parse(event.body));
        case 'COPY':
            return done(null, JSON.parse(event.body));
        case 'HEAD':
            return done(null, {});
        case 'OPTIONS':
            return done(null, JSON.parse(event.body));
        default:
            return done(new Error(`Unsupported method "${event.httpMethod}"`));
    }
};
